Simplify `JSON` ->`UTxO` conversion & add `Write.Tx` module

Anviking opened the pull request:
- [x] Simplify conversion from JSON/`[ApiExternalInput]`  to `UTxO` 
    - **Previously**: `[ApiExternalInput]` -> `[(W.TxIn, W.TxOut, Maybe (Hash "Datum"))]` -> `Cardano.UTxO era`
        - `W.TxOut` is incomplete, so we currently need the separate `Maybe (Hash "Datum")` and to re-combine it in `TransactionLayer.toCardanoUTxO`. This would get even more complicated with Babbage features.
        - `TransactionLayer.toCardanoUTxO` is partial and may call `error`
    - **Now**: `[ApiExternalInput]` -> `[(WriteTx.TxIn, WriteTx.TxOut)]` -> `Ledger.UTxO era` -> `Ledger.UTxO era`
        - where `WriteTx.TxOut` is isomorphic to `Ledger.TxOut BabbageEra`
        - More graceful error handling with: `utxoFromTxOutsInRecentEra :: forall era. RecentEra era -> [(TxIn, TxOutInRecentEra)] -> Either ErrInvalidTxOutInEra (Shelley.UTxO (ShelleyLedgerEra era))`
- [x] Add `Cardano.Wallet.Write.Tx` module supporting the above
    - Uses cardano-api-style `RecentEra era` GADT and `IsRecentEra era` constraint.

### Api changes
- Dropped support for balancing txs in Shelley, Allegra and Mary. The `balance_tx_byron_not_supported` error has been replaced with `balance_tx_era_not_supported `.
- New errors `inline_scripts_not_supported_in_alonzo`, and `inline_datums_not_supported_in_alonzo`, currently unused.

### Motivation

Preparation for #3436 

### Comments

<!-- Additional comments, links, or screenshots to attach, if any. -->

### Issue Number

ADP-1655

<!-- Reference the Jira/GitHub issue that this PR relates to, and which requirements it tackles.
  Note: Jira issues of the form ADP- will be auto-linked. -->

Anviking committed:
Copy `fromShelleyBasedScript` from cardano-api

Anviking commented on lib/wallet/src/Cardano/Wallet/Write/Tx/Gen.hs:
```diff
@@ -0,0 +1,63 @@
+import Cardano.Ledger.Alonzo.Data
+    ( Data (..), dataToBinaryData)
+import Cardano.Wallet.Write.Tx
+    ( type Datum, type DatumHash, pattern Datum, pattern DatumHash, pattern NoDatum, datumHashFromBytes, BinaryData )
```
Not thrilled about the `pattern Datum, type Datum` + `LANGUAGE ExplicitNamespaces, PatternSynonyms` requirement... But the only other alternatives would be:
1. Export a `Datum` type without pre-applied `BabbageEra StandardCrypto`
2. Create our own `data Datum = Datum .. | DatumHash ... | NoDatum` and convert to/from `Alonzo.Datum`

🤔

Anviking replied:
The `type Datum` seems to break `stylish-haskell`, so I can't leave it as is at least.

> Export a Datum type without pre-applied BabbageEra StandardCrypto

Perhaps hiding the `BabbageEra StandardCrypto` doesn't actually buy us anything. At least not at the moment. To construct a `TxOut era` with a `Datum era'` we _need_ `era ~ era'`.

Anviking replied:
I went with defining the `Write.Tx` functions in terms of e.g. `Datum LatestLedgerEra`. In the future we may want them to operate on `IsRecentEra era. era`, but let's deal with that... in the future.

Anviking commented on lib/wallet/test/unit/Cardano/Wallet/Api/TypesSpec.hs:
```diff
@@ -1428,18 +1432,18 @@ instance FromJSON SchemaApiErrorCode where
 
 -- Dummy instances
-instance EncodeAddress ('Testnet 0) where
+instance {-# INCOHERENT #-} EncodeAddress ('Testnet 0) where
```
Should add note and link to a tech dept ticket regarding removing the orphan instances.

Anviking committed:
Add `Cardano.Api.Gen.shrinkScriptData`

Anviking committed:
Add Compatibility.Ledger.Convert instance for TxIn

Anviking committed:
Rework JSON -> UTxO conversion with Write.Tx module

Remove need for ExplicitNameSpaces in Write.Tx

Rework `TxOutInRecentEra` (remove need for `wrapTxOutInRecentEra`)

Plus some improvement of doc comments.

Test `Script LatestLedgerEra` <->`Cardano.ScriptInAnyLang` roundtrip

Fix `genDatum` and add shrinkers

Add notes and roundtrips

Remove untested & unused scriptFromBytes

Improve `Write.Tx` test coverage

fixup: whitespace

fixup fromExternalInput formatting

Co-authored-by: Jonathan Knowles <jonathan.knowles@iohk.io>

fixup: ApiTypes indentation

Co-authored-by: Jonathan Knowles <jonathan.knowles@iohk.io>

fixup whitespace

fixup: reformat genData

Rename isRecentEra -> toRecentEra

Reformat toRecentEra

Reformat shelleyBasedEraFromRecentEra

fixup parens: Tx module

Tweak golden formatting in TxSpec

Co-authored-by: Jonathan Knowles <jonathan.knowles@iohk.io>

Reformat InAnyRecentEra

Co-authored-by: Jonathan Knowles <jonathan.knowles@iohk.io>

reformat scriptToCardanoEnvelopeJSON

reformat scriptFromCardanoEnvelopeJSON

Shrink bytestrings using `quickcheck-instances`

Tweak and remove redundant tests

Add module header

Co-authored-by: Jonathan Knowles <jonathan.knowles@iohk.io>

fixup comment in Tx module

Anviking committed:
Remove fromCardanoTx{In, Out} from TransactionLayer

Now with the shelley-core merge, we can just import the functions directly rather than going through the `TransactionLayer`.

jonathanknowles commented on lib/wallet/src/Cardano/Api/Extra.hs:
```diff
@@ -56,3 +73,48 @@ asAnyShelleyBasedEra = \case
+-- Copied from cardano-api because it is not exported.
+fromShelleyBasedScript  :: ShelleyBasedEra era
+                        -> Ledger.Script (ShelleyLedgerEra era)
+                        -> ScriptInEra era
```
```suggestion
fromShelleyBasedScript
    :: ShelleyBasedEra era
    -> Ledger.Script (ShelleyLedgerEra era)
    -> ScriptInEra era
```

jonathanknowles commented on lib/wallet/src/Cardano/Api/Extra.hs:
```diff
@@ -56,3 +73,48 @@ asAnyShelleyBasedEra = \case
+        Alonzo.PlutusScript Alonzo.PlutusV2 s ->
+          ScriptInEra PlutusScriptV2InBabbage $
+          PlutusScript PlutusScriptV2 $
+          PlutusScriptSerialised s
```
Request: format the above function with 4-space indents.

jonathanknowles commented on lib/wallet/src/Cardano/Api/Gen.hs:
```diff
@@ -666,9 +666,48 @@ genScriptData =
+        | n <= 1    = []
+        | otherwise = [ BS.take (n `div` 2) bs, BS.drop (n `div` 2) bs ]
+      where
+        n = BS.length bs
```
Just curious, is there any advantage to defining this shrinker, instead of using a pre-defined one?

I see there is one defined in the `quickcheck-instances` package:
https://hackage.haskell.org/package/quickcheck-instances-0.3.28/docs/Test-QuickCheck-Instances-ByteString.html

Anviking replied:
I got it from another place in the wallet, but using the shrinker from `quickcheck-instances` sounds decent 👍

jonathanknowles commented on lib/wallet/src/Cardano/Api/Gen.hs:
```diff
@@ -666,9 +666,48 @@ genScriptData =
+      where
+        n = BS.length bs
+
+
```
```suggestion

```

jonathanknowles commented on lib/wallet/src/Cardano/Wallet/Api/Server.hs:
```diff
@@ -4366,7 +4406,7 @@ getWalletTip
 getWalletTip ti = makeApiBlockReferenceFromHeader ti . currentTip
 
-fromExternalInput :: ApiExternalInput n -> (TxIn, TxOut, Maybe (Hash "Datum"))
+fromExternalInput :: ApiExternalInput n -> (WriteTx.TxIn, WriteTx.TxOutInRecentEra)
```
```suggestion
fromExternalInput
    :: ApiExternalInput n
    -> (WriteTx.TxIn, WriteTx.TxOutInRecentEra)
```

jonathanknowles commented on lib/wallet/src/Cardano/Wallet/Api/Types.hs:
```diff
@@ -1092,9 +1094,24 @@ data ApiExternalInput (n :: NetworkDiscriminant) = ApiExternalInput
+    parseJSON = withText "DatumHash" $ \hex -> maybeToParser $ do
+            bytes <- parseHex hex
+            ApiT <$> WriteTx.datumHashFromBytes bytes
+      where
```
```suggestion
    parseJSON = withText "DatumHash" $ \hex -> maybeToParser $ do
        bytes <- parseHex hex
        ApiT <$> WriteTx.datumHashFromBytes bytes
      where
```

jonathanknowles commented on lib/wallet/src/Cardano/Wallet/Shelley/Compatibility/Ledger.hs:
```diff
@@ -108,6 +110,8 @@ import qualified Cardano.Wallet.Primitive.Types.TokenBundle as TokenBundle
 import qualified Data.Map.Strict as Map
 import qualified Data.Map.Strict.NonEmptyMap as NonEmptyMap
+import Data.Maybe
+    ( fromMaybe )
```
Request: move this out of the qualified imports section and into the section above.

https://input-output-hk.github.io/adrestia/code/Coding-Standards#stylish-haskell-is-used-to-format-grouped-imports--language-pragmas

jonathanknowles commented on lib/wallet/test/unit/Cardano/Wallet/Shelley/Compatibility/LedgerSpec.hs:
```diff
@@ -50,6 +52,7 @@ spec = describe "Cardano.Wallet.Shelley.Compatibility.LedgerSpec" $
         ledgerRoundtrip $ Proxy @TokenName
         ledgerRoundtrip $ Proxy @TokenPolicyId
         ledgerRoundtrip $ Proxy @TokenQuantity
+        ledgerRoundtrip $ Proxy @TxIn
```
👍🏻 👍🏻 👍🏻

jonathanknowles commented on lib/wallet/src/Cardano/Wallet/Write/Tx.hs:
```diff
@@ -0,0 +1,481 @@
+    recentEra :: RecentEra era
+
+-- | Return a proof that the wallet can create txs in this era, or @Nothing@.
+isRecentEra :: Cardano.CardanoEra era -> Maybe (RecentEra era)
```
Using the prefix `is` implies that the function might return a `Bool`.

Given that this function actually returns a `RecentEra era`, perhaps rename to `toRecentEra`?

Anviking replied:
Done

jonathanknowles commented on lib/wallet/src/Cardano/Wallet/Write/Tx.hs:
```diff
@@ -0,0 +1,481 @@
+isRecentEra Cardano.MaryEra = Nothing
+isRecentEra Cardano.AllegraEra = Nothing
+isRecentEra Cardano.ShelleyEra = Nothing
+isRecentEra Cardano.ByronEra = Nothing
```
Suggestion: use `LambdaCase` for shorter code.

```suggestion
isRecentEra = \case
    Cardano.BabbageEra = Just RecentEraBabbage
    Cardano.AlonzoEra = Just RecentEraAlonzo
    Cardano.MaryEra = Nothing
    Cardano.AllegraEra = Nothing
    Cardano.ShelleyEra = Nothing
    Cardano.ByronEra = Nothing
```

jonathanknowles commented on lib/wallet/src/Cardano/Wallet/Write/Tx.hs:
```diff
@@ -0,0 +1,481 @@
+
+shelleyBasedEraFromRecentEra :: RecentEra era -> Cardano.ShelleyBasedEra era
+shelleyBasedEraFromRecentEra RecentEraBabbage = Cardano.ShelleyBasedEraBabbage
+shelleyBasedEraFromRecentEra RecentEraAlonzo = Cardano.ShelleyBasedEraAlonzo
```
Suggestion: use `LambdaCase` for shorter code.

Anviking replied:
I'm a bit torn in this case — I think the above way reads simpler in a way — but sure, going with it.

jonathanknowles commented on lib/wallet/src/Cardano/Wallet/Write/Tx.hs:
```diff
@@ -0,0 +1,481 @@
+     InAnyRecentEra :: IsRecentEra era -- Provide class constraint
+                    => RecentEra era   -- and explicit value.
+                    -> thing era
+                    -> InAnyRecentEra thing
```
```suggestion
data InAnyRecentEra thing where
     InAnyRecentEra
         :: IsRecentEra era -- Provide class constraint
         => RecentEra era   -- and explicit value.
         -> thing era
         -> InAnyRecentEra thing
```

jonathanknowles commented on lib/wallet/src/Cardano/Wallet/Write/Tx.hs:
```diff
@@ -0,0 +1,481 @@
+
+scriptFromCardanoScriptInAnyLang
+    :: Cardano.ScriptInAnyLang
+    -> (Script LatestLedgerEra)
```
```suggestion
    -> Script LatestLedgerEra
```

jonathanknowles commented on lib/wallet/src/Cardano/Wallet/Write/Tx.hs:
```diff
@@ -0,0 +1,481 @@
+    . fromMaybe (error "all valid scripts should be valid in latest era")
+    . Cardano.toScriptInEra latestEra
+  where
+    latestEra = Cardano.BabbageEra
```
Question: are there round-trip tests for:
- `script{From,To}CardanoScriptInAnyLang`?
- `script{From,To}CardanoEnvelopeJSON`?

Anviking replied:
Now, yes:

https://github.com/input-output-hk/cardano-wallet/blob/d0a7ef77bba61dddb1928c918b911ef2521a5d78/lib/wallet/test/unit/Cardano/Wallet/Write/TxSpec.hs#L125-L139

jonathanknowles commented on lib/wallet/src/Cardano/Wallet/Write/Tx.hs:
```diff
@@ -0,0 +1,481 @@
+        -> Aeson.Value
+    scriptToJSON (Cardano.ScriptInAnyLang l s) = Aeson.toJSON
+            $ obtainScriptLangConstraint l
+            $ Cardano.serialiseToTextEnvelope Nothing s
```
```suggestion
    scriptToJSON (Cardano.ScriptInAnyLang l s) = Aeson.toJSON
        $ obtainScriptLangConstraint l
        $ Cardano.serialiseToTextEnvelope Nothing s
```

jonathanknowles commented on lib/wallet/src/Cardano/Wallet/Write/Tx.hs:
```diff
@@ -0,0 +1,481 @@
+        obtainScriptLangConstraint
+          :: Cardano.ScriptLanguage lang
+          -> (Cardano.IsScriptLanguage lang => a)
+          -> a
```
```suggestion
        obtainScriptLangConstraint
            :: Cardano.ScriptLanguage lang
            -> (Cardano.IsScriptLanguage lang => a)
            -> a
```

jonathanknowles commented on lib/wallet/src/Cardano/Wallet/Write/Tx.hs:
```diff
@@ -0,0 +1,481 @@
+    envelope <- Aeson.parseJSON v
+    case textEnvelopeToScript envelope of
+      Left textEnvErr -> fail $ Cardano.displayError textEnvErr
+      Right (Cardano.ScriptInAnyLang l s) -> pure $ Cardano.ScriptInAnyLang l s
```
```suggestion
    case textEnvelopeToScript envelope of
        Left textEnvErr ->
            fail $ Cardano.displayError textEnvErr
        Right (Cardano.ScriptInAnyLang l s) ->
            pure $ Cardano.ScriptInAnyLang l s
```

jonathanknowles commented on lib/wallet/src/Cardano/Wallet/Write/Tx.hs:
```diff
@@ -0,0 +1,481 @@
+    . Map.map (Cardano.toShelleyTxOut (shelleyBasedEra @era))
+    . unCardanoUTxO
+  where
+    unCardanoUTxO (Cardano.UTxO m) = m
```
Question: is there a round-trip test for `{from,to}CardandUTxO`?

Anviking replied:
Added with the following note:

https://github.com/input-output-hk/cardano-wallet/blob/c525c1049f3b0984a27d68c11b523c478004adc5/lib/wallet/src/Cardano/Wallet/Write/Tx.hs#L472-L478

jonathanknowles commented on lib/wallet/src/Cardano/Wallet/Write/Tx/Gen.hs:
```diff
@@ -0,0 +1,56 @@
+{-# LANGUAGE TypeFamilies #-}
+module Cardano.Wallet.Write.Tx.Gen
```
```suggestion
{-# LANGUAGE TypeFamilies #-}

-- |
-- Copyright: © 2022 IOHK
-- License: Apache-2.0
--
module Cardano.Wallet.Write.Tx.Gen
```

jonathanknowles commented on lib/wallet/src/Cardano/Wallet/Write/Tx/Gen.hs:
```diff
@@ -0,0 +1,56 @@
+    ( fromMaybe )
+import Ouroboros.Consensus.Cardano.Block
+    ( EraCrypto, StandardCrypto )
+import Test.QuickCheck
```
Request: use explicit imports here for `Test.QuickCheck`.

jonathanknowles commented on lib/wallet/src/Cardano/Wallet/Write/Tx/Gen.hs:
```diff
@@ -0,0 +1,56 @@
+                <*> listOf (gendata (n `div` 2)),
+              PV1.List <$> listOf (gendata (n `div` 2))
+            ]
+    gendata _ = oneof [PV1.I <$> arbitrary, PV1.B <$> genByteString]
```
```suggestion
    gendata n
        | n > 0 = oneof
            [ PV1.I <$> arbitrary
            , PV1.B <$> genByteString
            , PV1.Map
                <$> listOf ((,) <$> gendata (n `div` 2) <*> gendata (n `div` 2))
            , PV1.Constr
                <$> fmap fromIntegral (arbitrary :: Gen Integer)
                <*> listOf (gendata (n `div` 2))
            , PV1.List <$> listOf (gendata (n `div` 2))
            ]
    gendata _ = oneof [PV1.I <$> arbitrary, PV1.B <$> genByteString]
```

Anviking commented on lib/wallet/src/Cardano/Wallet/Write/Tx.hs:
```diff
@@ -0,0 +1,481 @@
+            (Cardano.PlutusScriptLanguage Cardano.PlutusScriptV1) -> f
+            (Cardano.PlutusScriptLanguage Cardano.PlutusScriptV2) -> f
+
+scriptFromCardanoEnvelopeJSON
```
```suggestion
-- | Indirectly roundtrip-tested by "Api.TypesSpec"
scriptFromCardanoEnvelopeJSON
```

Anviking replied:
Or, right, they aren't here, but they will be in #3436...

Anviking replied:
I'm adding the missing tests... will push soon...

I discovered this gotcha though:
```
-- | NOTE: The type 'Script LatestLedgerEra' cannot distinguish between
-- 'Cardano.SimpleScriptV1' and 'Cardano.SimpleScriptV2'. This means the
-- roundtrip
-- @
--     scriptToCardanoScriptInAnyLang . scriptFromCardanoScriptInAnyLang
-- @
-- will convert 'SimpleScriptV1' to 'SimpleScriptV2'. Because 'SimpleScriptV1'
-- is 'ShelleyEra'-specific, and 'ShelleyEra' is not a 'RecentEra', this should
-- not be a problem.
```

jonathanknowles commented on lib/wallet/test/unit/Cardano/Wallet/Api/TypesSpec.hs:
```diff
@@ -1758,7 +1766,8 @@ instance Arbitrary (ApiExternalInput n) where
         <*> arbitrary
-        <*> arbitrary
+        <*> liftArbitrary (liftArbitrary genDatumHash)
+
```
```suggestion
        <*> liftArbitrary (liftArbitrary genDatumHash)
```

Anviking commented on lib/wallet/test/unit/Cardano/Wallet/Write/TxSpec.hs:
```diff
@@ -0,0 +1,201 @@
+{-# LANGUAGE ExplicitNamespaces #-}
```
```suggestion
```

jonathanknowles commented on lib/wallet/test/unit/Cardano/Wallet/Write/TxSpec.hs:
```diff
@@ -0,0 +1,201 @@
+                 let f = datumHashFromBytes . datumHashToBytes
+                 f h === Just h
+
+
```
```suggestion

```

jonathanknowles commented on lib/wallet/test/unit/Cardano/Wallet/Write/TxSpec.hs:
```diff
@@ -0,0 +1,201 @@
+                        [ B "hello"
+                        , B "\133\150\SOH\222\183rg+\147>\243\rf`\150\DLE\201(\
+                            \\188\241\SYN\149\SUBR\244\184i\143\&4\193\252\128"
+                        , Map [(B "",Map [(B "",I 2000000)])]
```
```suggestion
                        , Map [(B "", Map [(B "", I 2000000)])]
```

jonathanknowles commented on specifications/api/swagger.yaml:
```diff
@@ -4538,16 +4538,38 @@ x-errBadRequest: &errBadRequest
     message:
       type: string
-      description: Balancing Byron transactions is not supported.
+      description: Balancing transactions is not supported in this era.
```
👍🏻

jonathanknowles commented on lib/wallet/test/unit/Cardano/Wallet/Api/TypesSpec.hs:
```diff
@@ -941,18 +945,18 @@ instance FromJSON SchemaApiErrorCode where
     encodeAddress = const "<addr>"
 
-instance DecodeAddress ('Testnet 0) where
+instance {-# INCOHERENT #-} DecodeAddress ('Testnet 0) where
```
I'm a bit curious: why do the "real" instances for `{Decode,Encode}Address` and `{Decode,Encode}SharedAddress` (defined in `Cardano.Wallet.Shelley.Compatibility`) cause tests to fail in `Api.TypesSpec`?

If we were to simply delete these dummy instances, how much work would it be to adjust the test suite so that it uses the real instances?

Anviking replied:
Created and linked to: https://input-output.atlassian.net/browse/ADP-2302

Btw, it seems `OVERLAPPING` was enough, which already made it to master through some other PR. But I added TODOs with links anyway.

jonathanknowles commented on lib/wallet/src/Cardano/Wallet/Write/Tx.hs:
```diff
@@ -0,0 +1,520 @@
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeApplications #-}
+module Cardano.Wallet.Write.Tx
```
```suggestion

-- |
-- Copyright: © 2022 IOHK
-- License: Apache-2.0
--
module Cardano.Wallet.Write.Tx
```

Anviking replied:
8a0f55faf49df678c09e02bf9dd96d215184c062

jonathanknowles commented on lib/wallet/src/Cardano/Wallet/Write/Tx/Gen.hs:
```diff
@@ -0,0 +1,105 @@
+
+-- Originally from https://github.com/input-output-hk/cardano-ledger/blob/c7c63dabdb215ebdaed8b63274965966f2bf408f/eras/alonzo/test-suite/src/Test/Cardano/Ledger/Alonzo/Serialisation/Generators.hs#L66-L79
+genData :: Gen (BinaryData era)
+genData = dataToBinaryData . Data <$> resize 5 (sized gendata)
```
@Anviking 

I'm a bit curious: why use `resize` to hard-wire the QC size parameter here?

If we specify `resize` here with a fixed constant `c`, it would presumably prevent the generator from generating values smaller than `c`, which could prevent us from discovering bugs that only occur with smaller values.

Would it be feasible to instead use `scale` here, perhaps with `div` or `mod`? Then at least we'd still be able to generate a range of different sizes (albeit a smaller range).

Alternatively, our most common pattern in other `Gen` modules seems to be:

- make `Gen` modules define generators that depend on the QC size parameter (whatever that may be)
- if we need to alter the size in some way, either:
    - use `scale` at the point where we define an `Arbitrary` instance, or
    - if we need more fine-grained control over a parameter for a specific property, either:
        - use a newtype wrapper that adjusts the size (such as `Scale{Div,Mod}`).
        - use `forAll` in combination with an ordinary size combinator (such as `scale`).

Anviking replied:
I had just copied this from the ledger, but good point 👍 

What do you think about `scale (min 7)`? Then we allow both smaller sizes as well as slightly larger than what was previously generated.

Anviking committed:
Rework structure in TxSpec

*Before*

```
Isomorphisms
  Datum <-> Cardano.ScriptData
    datumFromCardanoScriptData . datumToCardanoScriptData == id [?]
      +++ OK, passed 100 tests.
    datumToCardanoScriptData . datumFromCardanoScriptData == id [?]
  Alonzo.BinaryData <-> Alonzo.Data
    dataToBinaryData . binaryDataToData == id [?]
      +++ OK, passed 100 tests.
    binaryDataToData . dataToBinaryData  == id [?]
      +++ OK, passed 100 tests.
  Alonzo.Data <-> Cardano.ScriptData
    Cardano.toAlonzoData . Cardano.fromAlonzoData == id [?]
      +++ OK, passed 100 tests.
    Cardano.fromAlonzoData . Cardano.toAlonzoData == id [?]
      +++ OK, passed 100 tests.
  Script LatestLedgerEra <-> Cardano.ScriptInAnyLang
    scriptToCardanoScriptInAnyLang . scriptFromCardanoScriptInAnyLang [?]
      +++ OK, passed 100 tests.
    scriptFromCardanoScriptInAnyLang . scriptToCardanoScriptInAnyLang [?]
      +++ OK, passed 100 tests.
  UTxO <-> Cardano.UTxO
    toCardanoUTxO . fromCardanoUTxO == id (modulo SimpleScriptV1/2) [?]
      +++ OK, passed 100 tests.
    fromCardanoUTxO . toCardanoUTxO == id [?]
      +++ OK, passed 100 tests.
Roundtrips
  datumFromBytes . datumToBytes == Right [?]
    +++ OK, passed 100 tests.
  datumHashFromBytes . datumHashToBytes == Just [?]
    +++ OK, passed 100 tests.
  parseEither (scriptFromCardanoEnvelopeJSON . scriptToCardanoEnvelopeJSON) == Right [?]
    +++ OK, passed 100 tests.
Goldens
  datumFromBytes
    I 42 [?]
    Constr 0 [] [?]
    Constr 1 [B "hello", B ..., Map ...] [?]
  scriptFromCardanoEnvelopeJSON [?]
```

Now:

```
Datum
  BinaryData is isomorphic to Cardano.ScriptData [?]
    +++ OK, passed 100 tests.
  datumFromBytes . datumToBytes == Right [?]
    +++ OK, passed 100 tests.
  datumHashFromBytes . datumHashToBytes == Just [?]
    +++ OK, passed 100 tests.
  datumFromBytes goldens
    I 42 [?]
    Constr 0 [] [?]
    Constr 1 [B "hello", B ..., Map ...] [?]
Script
  is isomorphic to Cardano.ScriptInAnyLang (modulo SimpleScriptV1/2) [?]
    +++ OK, passed 100 tests.
  parseEither (scriptFromCardanoEnvelopeJSON . scriptToCardanoEnvelopeJSON) == Right [?]
    +++ OK, passed 100 tests.
  scriptFromCardanoEnvelopeJSON golden [?]
UTxO
  is isomorphic to Cardano.UTxO (modulo SimpleScriptV1/2) [?]
    +++ OK, passed 100 tests.
```

Anviking committed:
Replace resize of genData

In response to https://github.com/input-output-hk/cardano-wallet/pull/3525#discussion_r997704917

Anviking committed:
fixup: actually use genData

Anviking committed:
Add DatumHash goldens & tweak API error message

Anviking committed:
fixup redundant import

Anviking committed:
Add TODOs re dummy {En,De}codeAddress

Anviking committed:
Extend Write.Tx module documentation

jonathanknowles commented on lib/wallet/src/Cardano/Wallet/Write/Tx/Gen.hs:
```diff
@@ -0,0 +1,105 @@
+
+-- Originally from https://github.com/input-output-hk/cardano-ledger/blob/c7c63dabdb215ebdaed8b63274965966f2bf408f/eras/alonzo/test-suite/src/Test/Cardano/Ledger/Alonzo/Serialisation/Generators.hs#L66-L79
+genData :: Gen (BinaryData era)
+genData = dataToBinaryData . Data <$> scale (min 7) (sized gendata)
```
I think the use of `min` is okay, but I'm curious as to the choice of `min` over `div` here?

Using `div` would have the advantage that the average size of `BinaryData` values would grow in proportion to the average size of `Datum` values, whereas `min` would impose a hard limit that breaks proportionality.

The `min` and `div` functions, when used with appropriate constants, give rise to similar distributions:
```hs
> fmap (sortOn fst . counts) $ generate $ replicateM 1000000 $ scale (`div` 4) $ arbitrarySizedNatural @Int 
[(0,124699),(1,125283),(2,124331),(3,124883),(4,125030),(5,125504),(6,125260),(7,125010)]

> fmap (sortOn fst . counts) $ generate $ replicateM 1000000 $ scale (`min` 7) $ arbitrarySizedNatural @Int 
[(0,125192),(1,124318),(2,124745),(3,124838),(4,125433),(5,125536),(6,125121),(7,124817)]
```

Anviking replied:
After some consideration and experimentation, I think it might make sense.

I was concerned over too large `BinaryData` values accidentally being generated (including by scaling in wrapping types), and was also not sure what the domain of sizes would be if we used `div`. 

But the sizes seem to range from 0 to 99

```haskell
quickCheck $ withMaxSuccess 10000 $ forAll getSize $ \s -> tabulate "size" [show s] $ property True

+++ OK, passed 10000 tests.

size (10000 in total):
 5.17% 0
 ...
 0.03% 99
```

where `scale (min 6)` would approximate `scale (const 6)`. So I guess something like `flip div 10` would allow us to have smaller values on average, without sacrificing extreme values.

Anviking replied:
430a26a89d3f45835f296b32c27c1c20e09b3f36

Anviking replied:
I'm not sure a `scale (2 *)` of a wrapping datatype (e.g. TxOut) would actually want to affect `genBinaryData` in its current form... but that's a problem for if/then.

jonathanknowles commented on lib/wallet/src/Cardano/Wallet/Write/Tx/Gen.hs:
```diff
@@ -0,0 +1,105 @@
+    ]
+
+-- Originally from https://github.com/input-output-hk/cardano-ledger/blob/c7c63dabdb215ebdaed8b63274965966f2bf408f/eras/alonzo/test-suite/src/Test/Cardano/Ledger/Alonzo/Serialisation/Generators.hs#L66-L79
+genData :: Gen (BinaryData era)
```
**Suggestion**: make the name of the generator reflect the returned type.

Justification: this follows the existing idiom of our `Gen` modules.

```suggestion
genBinaryData :: Gen (BinaryData era)
```

jonathanknowles commented on lib/wallet/src/Cardano/Wallet/Write/Tx/Gen.hs:
```diff
@@ -0,0 +1,105 @@
+shrinkDatum (DatumHash _) = [NoDatum]
+shrinkDatum NoDatum = []
+
+shrinkData :: BinaryData LatestLedgerEra -> [BinaryData LatestLedgerEra]
```
```suggestion
shrinkBinaryData :: BinaryData LatestLedgerEra -> [BinaryData LatestLedgerEra]
```

jonathanknowles commented on lib/wallet/test/unit/Cardano/Wallet/ApiSpec.hs:
```diff
@@ -138,7 +138,7 @@ import qualified Data.Map.Strict as Map
 import qualified Servant
 
-
+-- TODO [ADP-2302] Remove need for overlapping dummy instances
```
👍🏻

jonathanknowles commented on lib/wallet/api/http/Cardano/Wallet/Api/Http/Server/Error.hs:
```diff
@@ -381,11 +383,24 @@ instance IsServerError ErrDecodeTx where
-                "Balancing Byron transactions is not supported."
+        ErrOldEraNotSupported (Cardano.AnyCardanoEra era) ->
+            apiError err403 BalanceTxEraNotSupported $ T.unwords
+                [ "Balancing in ", T.pack (show era), " "
```
```suggestion
                [ "Balancing in ", showT era, " "
```

jonathanknowles commented on lib/wallet/api/http/Cardano/Wallet/Api/Http/Server/Error.hs:
```diff
@@ -381,11 +383,24 @@ instance IsServerError ErrDecodeTx where
+         WriteTx.ErrInlineDatumNotSupportedInAlonzo ->
+             apiError err400 BalanceTxInlineDatumsNotSupportedInAlonzo $ mconcat
+                 [ "Inline datums are not supported in the Alonzo era."
+                 ]
```
```suggestion
             apiError err400 BalanceTxInlineDatumsNotSupportedInAlonzo
                 "Inline datums are not supported in the Alonzo era."
```

jonathanknowles commented on lib/wallet/api/http/Cardano/Wallet/Api/Http/Server/Error.hs:
```diff
@@ -381,11 +383,24 @@ instance IsServerError ErrDecodeTx where
+         WriteTx.ErrInlineScriptNotSupportedInAlonzo ->
+             apiError err400 BalanceTxInlineScriptsNotSupportedInAlonzo $ mconcat
+                 [ "Inline scripts are not supported in the Alonzo era."
+                 ]
```
```suggestion
             apiError err400 BalanceTxInlineScriptsNotSupportedInAlonzo
                 "Inline scripts are not supported in the Alonzo era."
```

jonathanknowles commented on lib/wallet/api/http/Cardano/Wallet/Api/Types.hs:
```diff
@@ -1061,9 +1063,24 @@ data ApiExternalInput (n :: NetworkDiscriminant) = ApiExternalInput
+        maybeToParser = maybe failWithHelp pure
+        failWithHelp = fail $ mconcat
+            [ "expected hex-encoded 32-byte datum hash"
+            ]
```
```suggestion
        failWithHelp = fail
            "expected hex-encoded 32-byte datum hash"
```

jonathanknowles commented on lib/wallet/src/Cardano/Api/Extra.hs:
```diff
@@ -56,3 +73,49 @@ asAnyShelleyBasedEra = \case
+                PlutusScriptSerialised s
+            Alonzo.PlutusScript Alonzo.PlutusV2 _ ->
+                error "fromShelleyBasedScript: \
+                      \PlutusV2 not supported in Alonzo era"
```
```suggestion
                error
                    "fromShelleyBasedScript: \
                    \PlutusV2 not supported in Alonzo era"
```

jonathanknowles commented on lib/wallet/src/Cardano/Api/Gen.hs:
```diff
@@ -666,9 +668,40 @@ genScriptData =
                 [ ScriptDataList <$> vectorOf k smallerGen
                 , ScriptDataMap <$> vectorOf k ((,) <$> smallerGen <*> smallerGen)
-                , ScriptDataConstructor <$> arbitrary <*> vectorOf k smallerGen
+                , ScriptDataConstructor <$> genConstructorIx <*> vectorOf k smallerGen
```
```suggestion
                , ScriptDataConstructor
                    <$> genConstructorIx
                    <*> vectorOf k smallerGen
```

jonathanknowles commented on lib/wallet/src/Cardano/Api/Gen.hs:
```diff
@@ -666,9 +668,40 @@ genScriptData =
 
+        genConstructorIx = oneof
+            [ choose (0, 10) -- more realistic values
+            , abs <$> arbitrary
```
Perhaps:
```suggestion
            , arbitrarySizedNatural
```

Anviking replied:
671a4aec21858983f8b7d5c2e749f086b48520b2

Anviking replied:
I thought `Arbitrary Integer` generated large values, but I see that it doesn't. I replaced it with `arbitrary @Word64` instead.

`arbitrarySizedNatural` does sound like a neat replacement for `choose <small numbers>` though (or went with a mixture of both).

jonathanknowles commented on lib/wallet/src/Cardano/Api/Gen.hs:
```diff
@@ -666,9 +668,40 @@ genScriptData =
+        ScriptDataBytes _ -> []
+        ScriptDataConstructor _ l -> l
+
+    shrinkTuple f (a, b) = map (,b) (f a) ++ map (a,) (f b)
```
```suggestion
    shrinkTuple :: (t -> [t]) -> (t, t) -> [(t, t)]
    shrinkTuple f = liftShrink2 f f
```

Example:

```hs
> shrinkTuple f = liftShrink2 f f
> shrinkTuple shrinkIntegral (10, 10)
[(0,10),(5,10),(8,10),(9,10),(10,0),(10,5),(10,8),(10,9)]
```

Anviking replied:
👍

jonathanknowles commented on lib/wallet/src/Cardano/Api/Gen.hs:
```diff
@@ -666,9 +668,40 @@ genScriptData =
+        [ ScriptDataConstructor n' l'
+        | n' <- n : shrink n
+        , l' <- l : shrinkList shrinkScriptData l
+        ]
```
Suggestion: use `liftShrink2`

```suggestion
      ScriptDataConstructor n l ->                                                                                                                                                    
          uncurry ScriptDataConstructor <$>                                                                                                                                           
          liftShrink2 shrink (shrinkList shrinkScriptData) (n, l) 
```

Anviking replied:
👍 thanks!

jonathanknowles commented on lib/wallet/src/Cardano/Wallet/Write/Tx.hs:
```diff
@@ -0,0 +1,541 @@
+-- Compatibility
+--------------------------------------------------------------------------------
+
+
```
```suggestion

```

jonathanknowles commented on lib/wallet/test/unit/Cardano/Wallet/Write/TxSpec.hs:
```diff
@@ -0,0 +1,358 @@
+      where
+        genTxInOutEntry = ((,)
+            <$> genTxIn
+            <*> Cardano.genTxOut Cardano.BabbageEra)
```
```suggestion
        genTxInOutEntry = (,)
            <$> genTxIn
            <*> Cardano.genTxOut Cardano.BabbageEra
```

jonathanknowles commented on lib/wallet/test/unit/Cardano/Wallet/Write/TxSpec.hs:
```diff
@@ -0,0 +1,358 @@
+    shrink = shrink
+
+--------------------------------------------------------------------------------
+-- Workaround the distinction between SimpleScriptV1 and SimpleScriptV2 in
```
```suggestion
-- Work around the distinction between SimpleScriptV1 and SimpleScriptV2 in
```

(verb versus noun)

jonathanknowles commented on lib/wallet/test/unit/Cardano/Wallet/Write/TxSpec.hs:
```diff
@@ -0,0 +1,358 @@
+
+--------------------------------------------------------------------------------
+-- Workaround the distinction between SimpleScriptV1 and SimpleScriptV2 in
+-- cardano-api which neither we nor the ledger cares about.
```
```suggestion
-- cardano-api which neither we nor the ledger care about.
```

jonathanknowles reviewed (commented):
Hi @Anviking 

I'm still looking through, but here are some comments.

Anviking committed:
Rename to {gen,shrink}BinaryData

Anviking committed:
Tweak genConstructorIx

Anviking committed:
Use liftShrink2 in shrinkScriptData

Anviking committed:
Reformat genScriptData

Anviking committed:
Apply review suggestions

- Code style
- Grammar fixups

Co-authored-by: Jonathan Knowles <jonathan.knowles@iohk.io>

Anviking committed:
Rename datum{From,To}Bytes -> binaryData{To,From}Bytes

Anviking committed:
Reorganize export list

Anviking committed:
Fixup whitespace

Co-authored-by: Jonathan Knowles <jonathan.knowles@iohk.io>

Anviking committed:
fixup hlint

Anviking committed:
fixup: binaryDataToBytes renaming

Anviking committed:
Switch from `min` to `div` when scaling genBinaryData

If the size ranges from 0 to 99, this allow us to have smaller
values on average, without sacrificing extreme values.
